src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests_TestMarkup.cs

```diff
@@ -0,0 +1,404 @@
+       params FormattedClassification[] expected)
+    {
+        var allCode = $$"""""
+            class C
+            {
+                void M()
+                {
+                    Test.M(""""
+            {{code}}
+            """");
+                }
+            }
+            """"" + s_testMarkup;
+
+        var start = allCode.IndexOf(code, StringComparison.Ordinal);
+        var length = code.Length;
+        var span = new TextSpan(start, length);
+
+        var actual = await GetClassificationSpansAsync(allCode, span, options: null, testHost);
+
+        // Massage the results a bit so that the TestCode segments don't overlap the non-test-code segments.
+
+        var nonTestCodeSpans = actual.Where(s => s.ClassificationType != ClassificationTypeNames.TestCode).OrderBy((t1, t2) => t1.TextSpan.Start - t2.TextSpan.Start).ToImmutableArray();
+        var testCodeSpans = actual.Where(s => s.ClassificationType == ClassificationTypeNames.TestCode).OrderBy((t1, t2) => t1.TextSpan.Start - t2.TextSpan.Start).ToImmutableArray();
+
+        using var _ = ArrayBuilder<ClassifiedSpan>.GetInstance(out var totalSpans);
+
+        var normalizedNonTestCodeSpans = new NormalizedTextSpanCollection(nonTestCodeSpans.Select(c => c.TextSpan));
+        totalSpans.AddRange(nonTestCodeSpans);
+        foreach (var testCodeSpan in testCodeSpans)
+        {
+            var remainder = NormalizedTextSpanCollection.Difference(new NormalizedTextSpanCollection(testCodeSpan.TextSpan), normalizedNonTestCodeSpans);
+            foreach (var current in remainder)
+            {
+                if (!current.IsEmpty)
+                    totalSpans.Add(new ClassifiedSpan(current, testCodeSpan.ClassificationType));
+            }
+        }
+
+        var actualOrdered = totalSpans.OrderBy(static (t1, t2) => t1.TextSpan.Start - t2.TextSpan.Start).ToImmutableArray();
+        var actualFormatted = actualOrdered.SelectAsArray(a => new FormattedClassification(allCode.Substring(a.TextSpan.Start, a.TextSpan.Length), a.ClassificationType));
+        AssertEx.Equal(expected, actualFormatted);
+    }
+
+    [Theory, CombinatorialData]
+    public async Task TestEmbeddedCSharpMarkup1(TestHost testHost)
+    {
+        await TestEmbeddedCSharpAsync("""
+            class D
+            {
+            }
+            """,
+            testHost,
+            Keyword("class"),
+            TestCode(" "),
+            Class("D"),
+            Punctuation.OpenCurly,
+            Punctuation.CloseCurly);
+    }
+
+    [Theory, CombinatorialData]
+    public async Task TestEmbeddedCSharpCaret1(TestHost testHost)
+    {
+        await TestEmbeddedCSharpAsync("""
+            class D
+            {
+                $$
+            }
+            """,
+            testHost,
+            Keyword("class"),
+            TestCode(" "),
+            Class("D"),
+            Punctuation.OpenCurly,
+            TestCode("    "),
+            TestCodeMarkdown("$$"),
```

Comment:
TestCodeMarkdown refers to things like `$$` and `[|`.  The special markdown language we use for test files.